### ISSUE TITLE
Why does this SQL fail to hit the projection？

### ISSUE THREAD

**Issue by shadowDy** [ClickHouse/ClickHouse#36780]
add projection ：
`PROJECTION agg
    (
        SELECT
            a,
            b,
            c,
            floor(timestamp / 60000) AS tm,
            avg(used) / avg(quota) AS avg
        GROUP BY
            a,
            b,
            c,
            tm
    )`

when i execute sql：
`explain select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss') group by a,b,c,tm`
return： `ReadFromStorage (MergeTree(with 1 projection agg)`

when i execute sql：
`explain select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss') and a in ('aaa') and c in ('ccc')  group by a,b,c,tm`
return： `ReadFromStorage (MergeTree(with 1 projection agg)`

but 
case 1：when i execute sql (exclude a condition)：
`explain select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss')  and c in ('ccc')  group by a,b,c,tm`
return `SettingQuotaAndLimits (Set limits and quota after reading from storage) │
│                           ReadFromMergeTree`

case 2：when i execute sql (add timestamp condition)：
`explain select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss')  and c in ('ccc')  and timestamp between 123 and 456 group by a,b,c,tm`
return `SettingQuotaAndLimits (Set limits and quota after reading from storage) │
│                           ReadFromMergeTree`

Do I have to use TM here？

i expect agg can hit sql like:
`select a,b,c avg(if avg > 10 ,avg,null) avg from (
   select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss')  and c in ('ccc')  and timestamp between 123 and 456 group by a,b,c,tm)
group by a,b,c
`
@amosbird  How should I define it；thanks~

version: 21.9.5.16

**Comment by amosbird** [comment]
> Do I have to use TM here？

Yes, aggregate projection only recognize key columns in GROUP BY clause as a whole.

if  `where b in ('sss') and c in ('ccc')` doesn't use projection, it's likely that matched projection doesn't reduce the number of rows to read.

**Comment by shadowDy** [comment]
`select a,b,c avg(if avg > 10 ,avg,null) avg from ( select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss') and c in ('ccc') and timestamp between 123 and 456 group by a,b,c,tm) group by a,b,c`

I hope this query can be used in projection. Do you have any suggestions; thanks~
@amosbird

**Comment by shadowDy** [comment]
> Yes, aggregate projection only recognize key columns in GROUP BY clause as a whole.

Through the test, it is found that if you replace it with the following SQL, you can hit the target; It looks strange
`explain select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss') and c in ('ccc') and tm between 123/60000 and 456/60000 group by a,b,c,tm`
 or 
`explain select a,b,c,floor(timestamp / 60000) AS tm,avg(x) / avg(y) AS avg from test_table where b in ('sss') and c in ('ccc') and tm/60000 between 123and 456group by a,b,c,tm`

This should be optimized ~
